### PR TITLE
[Bugfix][DeepSeekV4] Skip MTP quant config for BF16 draft weights

### PR DESCRIPTION
**Pull request by yangsiqt**
## Summary

This PR fixes DeepSeek-V4 MTP draft model construction for artifacts whose `mtp.*` weights are BF16 on disk.

Some DeepSeek-V4 artifacts exclude the MTP block from quantization calibration. In that case, the target model is quantized, but the MTP draft block is stored as BF16 and does not have MTP-specific quantization scale tensors in the safetensors index.

Before this change, the MTP draft tower always reused the main model's `quant_config`. That makes the draft model allocate quantized parameter slots for the MTP block, while the checkpoint provides BF16 `mtp.*` weights. The loader then fails when mapping the on-disk weights to quantized parameter names.

This PR detects whether the MTP block has quantization tensors on disk. If no `mtp.*` quantization scale / packed-weight keys are found, the draft tower is constructed without `quant_config` so that its parameter structure matches the BF16 checkpoint.

Related PR:

- #43892 fixes an earlier DeepSeek-V4 optional-scale loading issue for the same artifact. The code change in this PR is independent, but the end-to-end validation below was run together with #43892 because current `main` otherwise fails before reaching the MTP draft model path.

## Scope

This PR is limited to DeepSeek-V4 MTP draft model construction.

It does not change the target model quantization path.

It does not attempt to fix DeepGEMM `next_n=3` support. For this PR, the primary validation target is `num_speculative_tokens=1`, which maps to the existing DeepGEMM `next_n=2` path.

## Implementation

The helper `_mtp_block_is_quantized_on_disk()` reads `model.safetensors.index.json` from the draft model path and scans only `mtp.*` keys.

It returns `True` if it finds any MTP quantization-related suffix, such as:

```text
.weight_scale
.weight_scale_inv
.weight_packed
.weight_global_scale
.input_global_scale
.weight_zero_point
```

If no such key is found for `mtp.*`, the MTP block is treated as BF16 on disk and the draft tower uses:

```python
quant_config = None
```

The fallback is conservative: if the index is missing or cannot be read, the code keeps the current upstream behavior and assumes the MTP block is quantized.

## Validation

Tested on:

```text
GPU: 2x NVIDIA H20
vLLM base: current main + #43892
Model: DeepSeek-V4-Flash-W4A16-FP8-MTP
TP: 2
max_model_len: 4096
kv_cache_dtype: fp8
safetensors_load_strategy: prefetch
```

Server command:

```bash
vllm serve "$MODEL" \
  --host 127.0.0.1 \
  --port 8000 \
  --tensor-parallel-size 2 \
  --kv-cache-dtype fp8 \
  --dtype auto \
  --gpu-memory-utilization 0.92 \
  --max-model-len 4096 \
  --max-num-seqs 4 \
  --max-num-batched-tokens 4096 \
  --safetensors-load-strategy prefetch \
  --speculative-config '{"method":"mtp","num_speculative_tokens":1}'
```

Observed log:

```text
MTP block weights are BF16 on disk (no quant scales for mtp.* keys in the safetensors index).
Constructing the MTP draft tower without quantization to match the artifact shape.
MTP draft model loaded: 31 params
```

Result:

```text
- MTP draft model loads successfully
- server reaches Application startup complete
- /v1/completions returns 200 OK
- spec_decode metrics are emitted
```

Smoke request:

```bash
curl http://127.0.0.1:8000/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "'"$MODEL"'",
    "prompt": "Write one short sentence about speculative decoding.",
    "max_tokens": 16,
    "temperature": 0
  }'
```

Observed response:

```text
HTTP 200 OK
completion_tokens: 16
```

Also checked:

```bash
python -m py_compile vllm/models/deepseek_v4/nvidia/mtp.py
```

## Note

I also verified `num_speculative_tokens=2` locally with a patched DeepGEMM build that supports SM90 paged MQA logits `next_n=3`, but that DeepGEMM change is separate from this vLLM-side fix. This PR only addresses the MTP draft model quantization mismatch.

### CODE REVIEW

**Comment by benchislett** [comment]
Is this the model you are testing with?
https://huggingface.co/LordNeel/DeepSeek-V4-Flash-Acti-MTP-W4A16-FP8

I don't see why the HF model can't just configure the quant config properly, as NVIDIA does with their NVFP4 checkpoint: https://huggingface.co/nvidia/DeepSeek-V4-Pro-NVFP4/blob/main/hf_quant_config.json

**Comment by yangsiqt** [comment]
> Is this the model you are testing with? https://huggingface.co/LordNeel/DeepSeek-V4-Flash-Acti-MTP-W4A16-FP8
> 
> I don't see why the HF model can't just configure the quant config properly, as NVIDIA does with their NVFP4 checkpoint: https://huggingface.co/nvidia/DeepSeek-V4-Pro-NVFP4/blob/main/hf_quant_config.json

Thanks for @benchislett taking a look.

No, the model I validated with is not `LordNeel/DeepSeek-V4-Flash-Acti-MTP-W4A16-FP8`.

The artifact I tested is:

`canada-quant/DeepSeek-V4-Flash-W4A16-FP8-MTP`

In that artifact, the target model is quantized, but the MTP block is preserved on disk as BF16. I checked `model.safetensors.index.json`: there are `mtp.*` tensors, but no MTP quantization companion tensors such as `.weight_scale`, `.weight_scale_inv`, `.weight_packed`, `.weight_global_scale`, etc. So if vLLM reuses the main model `quant_config` for the MTP draft tower, it constructs quantized MTP parameters while the checkpoint provides BF16 MTP weights, and loading fails.

I agree that the cleanest solution would be for the HF artifact to declare the quantization config precisely, similar to the NVIDIA checkpoint you linked. My intent with this PR was to make the vLLM DeepSeek-V4 MTP loader robust to existing artifacts where the main model is quantized but the MTP block is intentionally unquantized/BF16.

If you prefer this to be handled strictly as a model packaging issue, I can adjust the PR accordingly. Alternatively, I can narrow the implementation to rely on an explicit quant config / ignore entry instead of detecting the MTP quantization state from the safetensors index.

**Comment by benchislett** [comment]
I think there is already an established pattern here and we should expect independent quantizers to properly configure their quant_configs accordingly. Please file an issue on the HF model repo if you want support.
